onchain/src/art_peace.cairo

```diff
@@ -976,5 +984,107 @@ pub mod ArtPeace {
+
+        let mut max_scores: Felt252Dict<u32> = Default::default();
+        let mut max_indexes: Felt252Dict<u8> = Default::default();
+        let mut max_index: u8 = 0;
+
+        let mut votable_index: u8 = 1; // 0 means no vote
+
+        while votable_index <= votable_colors_count {
+            max_index = 0;
+            let mut index = votable_index;
+            let mut vote = self.color_votes.read((index, day));
+            // update max scores if needed
+            while max_index < nb_max {
+                let mut m: felt252 = max_index.into();
+                if (max_scores.get(m) < vote) {
+                    let mut old_vote = max_scores.get(m);
+                    let mut old_index = max_indexes.get(m);
+                    max_scores.insert(m, vote);
+                    max_indexes.insert(m, index);
+                    vote = old_vote;
+                    index = old_index;
+
+                    max_index += 1;
+                    // update max dict
+                    while max_index < nb_max {
+                        m = max_index.into();
+                        old_vote = max_scores.get(m);
+                        old_index = max_indexes.get(m);
+                        max_scores.insert(m, vote);
+                        max_indexes.insert(m, index);
+                        vote = old_vote;
+                        index = old_index;
+                        max_index += 1;
+                    };
+                };
+                max_index += 1;
+            };
+            votable_index += 1;
+        };
+
+        // update palette
+        max_index = 0;
+        while max_index < nb_max {
+            if (max_scores.get(max_index.into()) == 0) {
+                // no more max
+                break;
+            }
+            let index: u8 = max_indexes.get(max_index.into());
+
+            let color = self.votable_colors.read((index, day));
+            let color_index = self.color_count.read();
+            self.color_palette.write(color_index, color);
+            self.color_count.write(color_index + 1);
+            max_index += 1;
+        };
+
+        // update votable colors
+        let mut next_day_votable_colors_index: Array<u8> = array![];
+        votable_index = 1; // index 0 is unreachable
+        while votable_index <= votable_colors_count {
+            let mut found: u8 = 0;
+            max_index = 0;
+            while max_index < nb_max {
+                if (max_scores.get(max_index.into()) == 0) {
+                    // no more max
+                    break;
+                }
+                let index: u8 = max_indexes.get(max_index.into());
+                if index == votable_index {
+                    found = 1;
+                    break;
+                };
+                max_index += 1;
+            };
+            if found == 0 {
+                next_day_votable_colors_index.append(votable_index);
```

Comment:
Could you just do the `self.votable_colors.write` here and have a separate `new_votable_colors_index` variable to track that?

Just to make it a bit more efficient